Changes to set readOnlyRootFilesystem to true

arun-kbharathan-e3768 opened the pull request:
### Description
This PR implements proper support for ReadOnlyRootFilesystem: true in the OpenSearchCluster CRD, enhancing container security without compromising functionality.

### Issues Resolved
Earlier the container goes to Crashloop when ReadOnlyRootFilesystem is set to true.

### Check List
- [X] Commits are signed per the DCO using --signoff 
- [X] Unittest added for the new/changed functionality and all unit tests are successful
- [X] Customer-visible features documented
- [X] No linter warnings (`make lint`)

If CRDs are changed:
- [X] CRD YAMLs updated (`make manifests`) and also copied into the helm chart
- [X] Changes to CRDs documented

Please refer to the [PR guidelines](https://github.com/opensearch-project/opensearch-k8s-operator/blob/main/docs/developing.md#submitting-a-pr) before submitting this pull request.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

arun-kbharathan-e3768 commented:
This fixes  #951

rootxrishabh commented:
@arun-kbharathan-e3768, can you please resolve conflicts?

josedev-union commented on docs/userguide/main.md:
```diff
@@ -1432,3 +1432,83 @@ spec:
+nodePools:
+  initContainers:
+    - name: init-copier
+      image: opensearchproject/opensearch:2.17.1
```
This fix resolves the issue by 
- mounting an `emptyDir` volume for `/usr/share/opensearch/config` before mounting `ConfigMap` or `Secret` volumes for sub paths within that directory. 
- using `initContainers`, it ensures that all required files from these sources are copied into the `emptyDir`, providing OpenSearch with a writable configuration directory at runtime.

I tested this PR with 3.1 opensearch cluster. This is the yaml manifest.

```yaml
apiVersion: opensearch.opster.io/v1
kind: OpenSearchCluster
metadata:
  annotations:
    meta.helm.sh/release-name: opensearch-cluster
    meta.helm.sh/release-namespace: default
  creationTimestamp: "2025-10-17T02:38:38Z"
  finalizers:
  - Opster
  generation: 2
  labels:
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/version: 2.7.0
    helm.sh/chart: opensearch-cluster-3.0.0
  name: opensearch-cluster
  namespace: default
  resourceVersion: "940728"
  uid: 1153b533-77c3-4032-a67e-b6b8ae736960
spec:
  bootstrap:
    affinity: {}
    initContainers:
    - command:
      - bash
      - -c
      - cp -r /usr/share/opensearch/plugins/* /plugins-tmp && cp -r /usr/share/opensearch/config/*
        /config-tmp
      image: opensearchproject/opensearch:3.1.0
      name: init-copier
      resources: {}
      volumeMounts:
      - mountPath: /config-tmp
        name: rw-config
      - mountPath: /plugins-tmp
        name: rw-plugins
    resources: {}
  confMgmt: {}
  dashboards:
    affinity: {}
    enable: true
    image: docker.io/opensearchproject/opensearch-dashboards:3.1.0
    imagePullPolicy: IfNotPresent
    opensearchCredentialsSecret: {}
    podSecurityContext: {}
    replicas: 1
    resources: {}
    securityContext: {}
    service:
      type: ClusterIP
    tls:
      caSecret: {}
      generate: true
      secret: {}
    version: 3.1.0
  general:
    additionalVolumes:
    - emptyDir: {}
      name: rw-tmp
      path: /tmp
    - emptyDir: {}
      name: rw-config
      path: /usr/share/opensearch/config
    - emptyDir: {}
      name: rw-plugins
      path: /usr/share/opensearch/plugins
    - emptyDir: {}
      name: rw-logs
      path: /usr/share/opensearch/logs
    drainDataNodes: true
    httpPort: 9200
    image: docker.io/opensearchproject/opensearch:3.1.0
    imagePullPolicy: IfNotPresent
    monitoring:
      scrapeInterval: 30s
      tlsConfig: {}
    podSecurityContext: {}
    securityContext:
      readOnlyRootFilesystem: true
    serviceName: opensearch-cluster
    setVMMaxMapCount: true
    vendor: Opensearch
    version: 3.1.0
  initHelper:
    imagePullPolicy: IfNotPresent
    resources: {}
    version: "1.36"
  nodePools:
  - component: masters
    diskSize: 30Gi
    initContainers:
    - command:
      - bash
      - -c
      - bin/opensearch-plugin -v install --batch repository-s3 && cp -r /usr/share/opensearch/plugins/*
        /plugins-tmp && cp -r /usr/share/opensearch/config/* /config-tmp
      image: opensearchproject/opensearch:3.1.0
      name: init-copier
      resources: {}
      volumeMounts:
      - mountPath: /config-tmp
        name: rw-config
      - mountPath: /plugins-tmp
        name: rw-plugins
    replicas: 3
    resources:
      limits:
        cpu: 500m
        memory: 2Gi
      requests:
        cpu: 500m
        memory: 2Gi
    roles:
    - master
    - data
  security:
    config:
      adminCredentialsSecret: {}
      adminSecret: {}
      securityConfigSecret: {}
      updateJob:
        resources: {}
    tls:
      http:
        caSecret: {}
        generate: true
        secret: {}
      transport:
        caSecret: {}
        generate: true
        perNode: true
        secret: {}
```
```sh
NAME                 HEALTH   NODES   VERSION   PHASE     AGE
opensearch-cluster   green    3       3.1.0     RUNNING   16m
```

josedev-union replied:
I tested this in 2.17 also. It works ok.

arun-kbharathan-e3768 commented:
> @arun-kbharathan-e3768 could you resolve the conflict? btw, LGTM. Once this is rebased, i will approve.

Thanks, Will update today

arun-kbharathan-e3768 commented:
> @arun-kbharathan-e3768 I opened a cloned PR #1143 with rebase. It would be great if you allow write access to your forked repo so other contributors can contribute to solve conflicts. If we don't get this conflict resolved soon, we will merge that new PR. (I added you as a co-author in the new PR btw.)

hi i added user @josedev-union as collaborator

synhershko commented:
@josedev-union I think we can close this?

josedev-union commented:
> @josedev-union I think we can close this?

yes
